src/main/java/by/itech/lab/supplier/service/impl/WarehouseServiceImpl.java

```diff
@@ -28,15 +47,70 @@ public WarehouseDto findById(Long warehouseId) {
                 .orElseThrow(() -> new ResourceNotFoundException("Warehouse with id=" + warehouseId + " doesn't exist"));
     }
 
-    // TODO: 11/4/20
+    @Transactional
     @Override
-    public WarehouseDto save(WarehouseDto dto) {
-        return null;
+    public WarehouseDto save(Long customerId, WarehouseDto warehouseDto) {
+        if (warehouseDto.getIdentifier() == null) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_ACCEPTABLE, "\"Identifier\" can't be null");
+        }
+
+        Warehouse warehouse = warehouseMapper.map(warehouseDto);
+        Address savedAddress = addressCreate(warehouseDto.getAddressDto());
+        warehouse.setAddress(savedAddress);
+        Optional<Customer> customer = customerRepository.findById(customerId);
+        customer.ifPresent(warehouse::setCustomer);
+        final Warehouse savedWarehouse = warehouseRepository.save(warehouse);
+
+        for (User user : warehouse.getUsers()) {
+            user.setWarehouse(warehouse);
+            userRepository.save(user);
+        }
+
+        return warehouseMapper.map(savedWarehouse);
     }
 
-    // TODO: 11/4/20
+    @Transactional
     @Override
-    public void delete(Long id) {
+    public WarehouseDto update(WarehouseDto warehouseDto) {
+        Warehouse entity = warehouseRepository.findById(warehouseDto.getId()).orElseThrow();
+        if (!warehouseDto.getIdentifier().equals(entity.getIdentifier())) {
+            throw new ResponseStatusException(
+                    HttpStatus.NOT_ACCEPTABLE, "You can't change \"identifier\" field");
+        }
+
+        final Address existingAddress = addressRepository.findById(entity.getAddress().getId()).orElseThrow();
+        warehouseDto.getAddressDto().setId(existingAddress.getId());
+        addressMapper.map(warehouseDto.getAddressDto(), existingAddress);
+        addressRepository.save(existingAddress);
 
+        Warehouse warehouse = Optional.ofNullable(warehouseDto.getId())
+                .map(item -> {
+                    final Warehouse existingWarehouse = warehouseRepository
+                            .findById(warehouseDto.getId())
+                            .orElseThrow();
+                    warehouseMapper.map(warehouseDto, existingWarehouse);
+                    return existingWarehouse;
+                })
+                .orElseThrow();
+
+        final Warehouse savedWarehouse = warehouseRepository.save(warehouse);
+        return warehouseMapper.map(savedWarehouse);
+    }
+
+    private Address addressCreate(AddressDto addressDto) {
+        Address address = addressMapper.map(addressDto);
+        return addressRepository.save(address);
+    }
+
+    @Transactional
+    @Override
+    public void delete(final Long id) {
+        warehouseRepository.delete(id);
+    }
+
+    @Override
+    public WarehouseDto save(WarehouseDto dto) {
+        return null;
     }
```

Comment:
If method does nothing, why do we need it?
You can put customer_id field into warehouseDto

Reply:
I should override this method, cuz it's in BaseService, when I use my own save method.

Reply:
Soooo, maybe you should change your own method to make it override method in BaseService?